### ISSUE TITLE
Spaces added to generator expressions

### ISSUE THREAD

**Issue by kumekay** [cheshirekow/cmake_format#153]
Spaces may be added some to CMake generator expressions,

For example:

```
file(GENERATE OUTPUT ${build_dir}/ldgen_libraries.in CONTENT $<JOIN:${ldgen_libraries},\n>)
```

is formatted to:

```
file(GENERATE OUTPUT ${build_dir}/ldgen_libraries.in CONTENT $<JOIN:${ldgen_libraries}, \n>)
```
that is not a valid CMake.

Original line: https://github.com/espressif/esp-idf/blob/master/tools/cmake/ldgen.cmake#L36

Version: 0.6.3

Config:
```
# --------------------------
# General Formatting Options
# --------------------------
# How wide to allow formatted cmake files
line_width = 120

# How many spaces to tab for indent
tab_size = 4

# If an argument group contains more than this many sub-groups (parg or kwarg
# groups), then force it to a vertical layout.
max_subgroups_hwrap = 3

# If a positional argument group contains more than this many arguments, then
# force it to a vertical layout.
max_pargs_hwrap = 6

# If true, separate flow control names from their parentheses with a space
separate_ctrl_name_with_space = False

# If true, separate function names from parentheses with a space
separate_fn_name_with_space = False

# If a statement is wrapped to more than one line, than dangle the closing
# parenthesis on it's own line.
dangle_parens = False

# If the trailing parenthesis must be 'dangled' on it's on line, then align it
# to this reference: `prefix`: the start of the statement,  `prefix-indent`: the
# start of the statement, plus one indentation  level, `child`: align to the
# column of the arguments
dangle_align = 'prefix'

min_prefix_chars = 4

# If the statement spelling length (including space and parenthesis) is larger
# than the tab width by more than this amount, then force reject un-nested
# layouts.
max_prefix_chars = 10

# If a candidate layout is wrapped horizontally but it exceeds this many lines,
# then reject the layout.
max_lines_hwrap = 2

# What style line endings to use in the output.
line_ending = 'unix'

# Format command names consistently as 'lower' or 'upper' case
command_case = 'canonical'

# Format keywords consistently as 'lower' or 'upper' case
keyword_case = 'unchanged'

# Specify structure for custom cmake functions
additional_commands = {
  "pkg_find": {
    "kwargs": {
      "PKG": "*"
    }
  },
  "ExternalProject_Add_StepDependencies": {
    "pargs": "*"
  }
}

# A list of command names which should always be wrapped
always_wrap = []

# If true, the argument lists which are known to be sortable will be sorted
# lexicographicall
enable_sort = True

# If true, the parsers may infer whether or not an argument list is sortable
# (without annotation).
autosort = False

# If a comment line starts with at least this many consecutive hash characters,
# then don't lstrip() them off. This allows for lazy hash rulers where the first
# hash char is not separated by space
hashruler_min_length = 10

# By default, if cmake-format cannot successfully fit everything into the
# desired linewidth it will apply the last, most agressive attempt that it made.
# If this flag is True, however, cmake-format will print error, exit with non-
# zero status code, and write-out nothing
require_valid_layout = False

# A dictionary containing any per-command configuration overrides. Currently
# only `command_case` is supported.
per_command = {}

# A dictionary mapping layout nodes to a list of wrap decisions. See the
# documentation for more information.
layout_passes = {}


# --------------------------
# Comment Formatting Options
# --------------------------
# What character to use for bulleted lists
bullet_char = '*'

# What character to use as punctuation after numerals in an enumerated list
enum_char = '.'

# enable comment markup parsing and reflow
enable_markup = False

# If comment markup is enabled, don't reflow the first comment block in each
# listfile. Use this to preserve formatting of your copyright/license
# statements.
first_comment_is_literal = False

# If comment markup is enabled, don't reflow any comment block which matches
# this (regex) pattern. Default is `None` (disabled).
literal_comment_pattern = None

# Regular expression to match preformat fences in comments
# default=r'^\s*([`~]{3}[`~]*)(.*)$'
fence_pattern = '^\\s*([`~]{3}[`~]*)(.*)$'

# Regular expression to match rulers in comments
# default=r'^\s*[^\w\s]{3}.*[^\w\s]{3}$'
ruler_pattern = '^\\s*[^\\w\\s]{3}.*[^\\w\\s]{3}$'

# If true, then insert a space between the first hash char and remaining hash
# chars in a hash ruler, and normalize it's length to fill the column
canonicalize_hashrulers = True


# ---------------------------------
# Miscellaneous Options
# ---------------------------------
# If true, emit the unicode byte-order mark (BOM) at the start of the file
emit_byteorder_mark = False

# Specify the encoding of the input file. Defaults to utf-8.
input_encoding = 'utf-8'

# Specify the encoding of the output file. Defaults to utf-8. Note that cmake
# only claims to support utf-8 so be careful when using anything else
output_encoding = 'utf-8'
```

Of course it's easy to avoid just by quoting the expression, but it seems to be an issue.

**Comment by cheshirekow** [comment]
Hm... my brain can't comprehend why your example doesn't match [this pattern][1], but I can reproduce the error. I'll see if I can figure out what's going on.

[1]: https://github.com/cheshirekow/cmake_format/blob/master/cmake_format/lexer.py#L121

**Comment by cheshirekow** [comment]
Oh, nevermind. Stupid me... `\t\r\n` shouldn't be escaped in the string... they should be `trn`... let me see if that fixes it.

**Comment by cheshirekow** [comment]
This patch seems to do the trick. Will be included in the next release:
```
diff --git a/cmake_format/lexer.py b/cmake_format/lexer.py
index e779d46..1e61dcf 100644
--- a/cmake_format/lexer.py
+++ b/cmake_format/lexer.py
@@ -104,15 +104,15 @@ def tokenize(contents):
   legacy_pattern = "({})+".format(
       "|".join([
           # Make-style variable like $(MAKE)
-          r'\$\([^\$\(\)]+\)',
+          r'(\$\([^\$\(\)]+\))',
           # Quoted-substring
-          r'"[^"\\]*(?:\\.[^"\\]*)*"',
+          r'("[^"\\]*(?:\\.[^"\\]*)*")',
           # Any element except whitespace or one of '()#"\'
-          r'[^\s\(\)#"\\]',
+          r'([^\s\(\)#"\\])',
           # Escape sequences
           # https://cmake.org/cmake/help/v3.0/manual/
           #   cmake-language.7.html#grammar-token-escape_sequence
-          r'\\[\(\)#" \\\$@\^\t\r\n;]'
+          r'(\\[\(\)#" \\\$@^trn;])'
       ])
   )
 
@@ -121,11 +121,11 @@ def tokenize(contents):
   unquoted_pattern = "({})+".format(
       "|".join([
           # Any element except whitespace or one of '()#"\'
-          r'[^\s\(\)#"\\]',
+          r'([^\s\(\)#"\\])',
           # Escape sequences
           # https://cmake.org/cmake/help/v3.0/manual/
           #   cmake-language.7.html#grammar-token-escape_sequence
-          r'\\[\(\)#" \\\$@^trn;]'
+          r'(\\[\(\)#" \\\$@^trn;])'
       ])
   )
```

**Comment by cheshirekow** [comment]
Fixed I think in [v0.6.4][1]. Let me know if you see more instances. 

[1]: https://github.com/cheshirekow/cmake_format/blob/v0.6.4/cmake_format/lexer_tests.py#L100
